File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void countGoodNodes(TreeNode* root, int maxValue) {
        if (!root) return;
        if (root->val >= maxValue) {
            maxValue = root->val;
            max_counter_++;
        }
        countGoodNodes(root->left, maxValue);
        countGoodNodes(root->right, maxValue);
    }

    int goodNodes(TreeNode* root) {
        countGoodNodes(root, root->val);
        return max_counter_;
    }

private:
    int max_counter_ = 0;
};

File: Data Structures & Algorithms/depth-of-binary-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = [(root, 1)]
        max_depth = 0
        
        while stack and root:
            node_depth = stack.pop()
            max_depth = max(max_depth, node_depth[1])
            if node_depth[0].right: stack.append((node_depth[0].right, node_depth[1] + 1))
            if node_depth[0].left: stack.append((node_depth[0].left, node_depth[1] + 1))

        return max_depth